### ISSUE TITLE
gnt-job hangs on cluster init (ganeti 2.11.5 + xen 4.4.0 + jessie)

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1020]
Originally reported of Google Code with ID 964.

```
Hi,

I'm trying to configure ganeti node with xen 4.4 and ganeti 2.11.5 on Debian Jessie. After cluster initialization I have found that gnt-job hangs. 

root@node01:/# gnt-job list
 ID Status  Summary
  1 running CLUSTER_POST_INIT
  4 running CLUSTER_SET_PARAMS
  8 running CLUSTER_VERIFY_GROUP(bfbaacb2-6198-4495-bb10-b5a42bed56d0)

root@node01:/# gnt-job info 1
Job ID: 1
  Status: running
  Received:         2014-09-28 22:56:41.547590
  Processing start: 2014-09-28 22:56:41.579392 (delta 0.031802s)
  Processing end:   unknown (None)
  Total processing time: N/A
  Opcodes:
    OP_CLUSTER_POST_INIT
      Status: running
      Processing start: 2014-09-28 22:56:41.579392
      Execution start:  2014-09-28 22:56:41.597571
      No processing end time
      Input fields:
        comment: None
        debug_level: 0
        depends: None
        priority: 0
        reason: ['gnt:client:gnt-cluster', 'init', 1411937801540854016],['gnt:opcode:cluster_post_init', 'job=1;index=0', 1411937801547590000],['gnt:daemon:masterd:pickup:op_cluster_post_init', 'job=1;index=0', 1411937801557820160]
      No output data
      Execution log:

root@node01:/# gnt-job info 4
Job ID: 4
  Status: running
  Received:         2014-09-28 23:09:11.161563
  Processing start: 2014-09-28 23:09:11.187818 (delta 0.026255s)
  Processing end:   unknown (None)
  Total processing time: N/A
  Opcodes:
    OP_CLUSTER_SET_PARAMS
      Status: running
      Processing start: 2014-09-28 23:09:11.187818
      Execution start:  2014-09-28 23:09:11.206144
      No processing end time
      Input fields:
        beparams: {}
        comment: None
        debug_level: 0
        default_iallocator_params: {}
        depends: None
        disk_state: {}
        diskparams: {}
        dry_run: False
        force: False
        hv_state: {}
        hvparams: {'xen-pvm': {'xen_cmd': 'xl'}}
        ipolicy: {}
        nicparams: {}
        priority: 0
        reason: ['gnt:client:gnt-cluster', 'modify', 1411938551158664960],['gnt:opcode:cluster_set_params', 'job=4;index=0', 1411938551161563000],['gnt:daemon:masterd:pickup:op_cluster_set_params', 'job=4;index=0', 1411938551170553856]
      No output data
      Execution log:

root@node01:/# gnt-job info 8
Job ID: 8
  Status: running
  Received:         2014-09-28 23:12:33.764673
  Processing start: 2014-09-28 23:12:33.848414 (delta 0.083741s)
  Processing end:   unknown (None)
  Total processing time: N/A
  Opcodes:
    OP_CLUSTER_VERIFY_GROUP
      Status: running
      Processing start: 2014-09-28 23:12:33.848414
      Execution start:  2014-09-28 23:12:34.231172
      No processing end time
      Input fields:
        comment: Submitted by OP_CLUSTER_VERIFY
        debug_level: 0
        debug_simulate_errors: False
        depends: [7, []]
        error_codes: False
        group_name: bfbaacb2-6198-4495-bb10-b5a42bed56d0
        ignore_errors: 
        priority: 0
        reason: ['gnt:client:gnt-cluster', 'verify', 1411938753714611968],['gnt:opcode:cluster_verify', 'job=6;index=0', 1411938753717190000],['gnt:daemon:masterd:pickup:op_cluster_verify', 'job=6;index=0', 1411938753723702016],['gnt:client:gnt-cluster', 'verify', 1411938753714611968],['gnt:opcode:cluster_verify', 'job=6;index=0', 1411938753717190000],['gnt:daemon:masterd:pickup:op_cluster_verify', 'job=6;index=0', 1411938753723702016],['gnt:opcode:cluster_verify_group', 'job=8;index=0', 1411938753764673000],['gnt:daemon:masterd:pickup:op_cluster_verify_group', 'job=8;index=0', 1411938753829798912]
        skip_checks: 
        verbose: False
      No output data
      Execution log:
        1:2014-09-28 23:12:34.239590:message * Verifying group 'default'
        2:2014-09-28 23:12:34.247218:message * Gathering data (1 nodes)


After 2 days, It is still in running state. I have also try to restart gnt service and try to change xen-hvm parameter:

root@node01:/# gnt-cluster modify -H xen-pvm:xen_cmd=xl

But still no luck for me ...

More info:

What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

root@node01:/# gnt-cluster --version
gnt-cluster (ganeti v2.11.5) 2.11.5

root@node01:/# gnt-cluster version
Software version: 2.11.5
Internode protocol: 2110000
Configuration format: 2110000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.11.5

root@node01:/# hspace --version
hspace (ganeti) version v2.11.5
compiled with ghc 7.6
running on linux x86_64

root@node01:/# gnt-cluster info
Cluster name: cluster01.ganeti.lan
Cluster UUID: d5a6d5de-9536-48e5-bc5c-f672f74ab51b
Creation time: 2014-09-28 22:55:54
Modification time: 2014-09-28 22:55:54
Master node: node01.ganeti.lan
Architecture (this node): 64bits (x86_64)
Tags: (none)
Default hypervisor: xen-pvm
Enabled hypervisors: xen-pvm
Hypervisor parameters: 
  xen-pvm: 
    blockdev_prefix: sd
    bootloader_args: 
    bootloader_path: 
    cpu_cap: 0
    cpu_mask: all
    cpu_weight: 256
    cpuid: 
    initrd_path: 
    kernel_args: ro
    kernel_path: /boot/vmlinuz-3-xenU
    migration_mode: live
    migration_port: 8002
    reboot_behavior: reboot
    root_path: /dev/xvda1
    soundhw: 
    use_bootloader: False
    vif_script: 
    xen_cmd: xm
OS-specific hypervisor parameters: 
OS parameters: 
Hidden OSes: 
Blacklisted OSes: 
Cluster parameters: 
  candidate pool size: 10
  maximal number of jobs running simultaneously: 20
  master netdev: xenbr0
  master netmask: 32
  use external master IP address setup script: False
  lvm volume group: node01-vg
  lvm reserved volumes: (none)
  drbd usermode helper: /bin/true
  file storage path: /srv/ganeti/file-storage
  shared file storage path: /srv/ganeti/shared-file-storage
  gluster storage path: /var/run/ganeti/gluster
  maintenance of node health: False
  uid pool: 
  default instance allocator: hail
  default instance allocator parameters: 
  primary ip version: 4
  preallocation wipe disks: False
  OS search path: /srv/ganeti/os, /usr/local/lib/ganeti/os, /usr/lib/ganeti/os, /usr/share/ganeti/os
  ExtStorage Providers search path: /srv/ganeti/extstorage, /usr/local/lib/ganeti/extstorage, /usr/lib/ganeti/extstorage, /usr/share/ganeti/extstorage
  enabled disk templates: drbd, plain
  enabled user shutdown: False
Default node parameters: 
  exclusive_storage: False
  oob_program: 
  ovs: False
  ovs_link: 
  ovs_name: switch1
  spindle_count: 1
  ssh_port: 22
Default instance parameters: 
  default: 
    always_failover: False
    auto_balance: True
    maxmem: 128
    minmem: 128
    spindle_use: 1
    vcpus: 1
Default nic parameters: 
  default: 
    link: xen-br0
    mode: bridged
    vlan: 
Default disk parameters: 
  blockdev: 
  diskless: 
  drbd: 
    c-delay-target: 1
    c-fill-target: 0
    c-max-rate: 61440
    c-min-rate: 4096
    c-plan-ahead: 20
    data-stripes: 1
    disk-barriers: n
    disk-custom: 
    dynamic-resync: False
    meta-barriers: False
    meta-stripes: 1
    metavg: node01-vg
    net-custom: 
    protocol: C
    resync-rate: 61440
  ext: 
  file: 
  gluster: 
    access: kernelspace
    host: 127.0.0.1
    port: 24007
    volume: gv0
  plain: 
    stripes: 1
  rbd: 
    access: kernelspace
    pool: rbd
  sharedfile: 
Instance policy - limits for instances: 
  bounds specs: 
    - max/0: 
        cpu-count: 8
        disk-count: 16
        disk-size: 1048576
        memory-size: 32768
        nic-count: 8
        spindle-use: 12
      min/0: 
        cpu-count: 1
        disk-count: 1
        disk-size: 1024
        memory-size: 128
        nic-count: 1
        spindle-use: 1
  std: 
    cpu-count: 1
    disk-count: 1
    disk-size: 1024
    memory-size: 128
    nic-count: 1
    spindle-use: 1
  allowed disk templates: drbd, plain
  vcpu-ratio: 4
  spindle-ratio: 32

<b>What distribution are you using?</b>

root@node01:/# uname -a
Linux node01 3.16-2-amd64 #1 SMP Debian 3.16.3-2 (2014-09-20) x86_64 GNU/Linux

root@node01:/# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux testing (jessie)
Release:        testing
Codename:       jessie

Xen info:

root@node01:/# dmesg | grep Xen\ version
[    0.000000] Xen version: 4.4.0 (preserve-AD)

root@node01:/# xl info
host                   : node01
release                : 3.16-2-amd64
version                : #1 SMP Debian 3.16.3-2 (2014-09-20)
machine                : x86_64
nr_cpus                : 4
max_cpu_id             : 3
nr_nodes               : 1
cores_per_socket       : 4
threads_per_core       : 1
cpu_mhz                : 1958
hw_caps                : 178bfbff:28100800:00000000:00000100:00000201:00000000:00000001:00000000
virt_caps              :
total_memory           : 3071
free_memory            : 128
sharing_freed_memory   : 0
sharing_used_memory    : 0
outstanding_claims     : 0
free_cpus              : 0
xen_major              : 4
xen_minor              : 4
xen_extra              : .0
xen_version            : 4.4.0
xen_caps               : xen-3.0-x86_64 xen-3.0-x86_32p 
xen_scheduler          : credit
xen_pagesize           : 4096
platform_params        : virt_start=0xffff800000000000
xen_changeset          : 
xen_commandline        : placeholder
cc_compiler            : gcc (Debian 4.9.1-11) 4.9.1
cc_compile_by          : waldi
cc_compile_domain      : debian.org
cc_compile_date        : Sat Sep  6 20:31:50 UTC 2014
xend_config_format     : 4

Thanks !
```
Originally added on 2014-09-30 16:22:10 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I tried reproducing this on a Wheezy node with a backported Xen 4.4, and all the post-init processes completed successfully, despite xm and not xl selected as the Xen tool.

Could you provide any more information gathered from e.g. logs? Look in /var/log/ganeti on the master node, the master logs and node daemon logs would be most useful. If there are other nodes, the node daemon log from those would be a good place to look as well.
```
Originally added on 2014-10-02 14:33:04 +0000 UTC.

Changed State: NeedFeedback

**Comment by Ganeti-Issues-Migrator** [comment]
```
that looks just like my #961

to check if it is the same error pls send us a 'ps aux | grep ganeti' - I guess luxid will be missing because it crashed

apollon had recomended to upgrade to 2.11.6 from unstable because there was some fix in there (with some curl-stuff(?))
but I haven't had time to test that yet, maybe I'll do so today evening
```
Originally added on 2014-10-02 16:25:25 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

"try installing libgnutls-deb0-28 and libgnutls-openssl27 by hand from here: http://snapshot.debian.org/package/gnutls28/3.2.15-2/"
And Upgrade to 2.11.6-1 (from sid)

That should solve this problem.
But maybe you will still have to re-init the cluster as the failed jobs are still "running"
```
Originally added on 2014-10-03 16:32:18 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 6e06de4e7efbb037b9a46b8c2063a05b2921d94f
Author: Apollon Oikonomopoulos <apoikos@gmail.com>
Date:   Sat Oct 4 21:14:03 2014 +0300

    ganeti.daemon: fix daemon mode with GnuTLS >= 3.3

    Newer GnuTLS versions (>= 3.3.0) use a library constructor for
    initialization and open /dev/urandom on library load, way before we
    fork(). Closing /dev/urandom on fork causes a failure to re-seed GnuTLS's
    random number generator during the first ganeti.http.client request, which
    in turn causes the process to silently abort(3).

    For more background on this behavior, see this thread at the GnuTLS
    mailing list:

    http://lists.gnupg.org/pipermail/gnutls-help/2014-April/003429.html

    Note that calling pycurl.global_init() at the correct place (as we do) is not
    enough, as it does not cause a re-initialization of the GnuTLS library.

    As we cannot reliably detect neither the GnuTLS version, nor the socket, we
    work our way around this by keeping all fds referring to /dev/urandom open
    after fork. We do so using the /proc/self/fd interface.

    This fixes issues #961 and #964.

    Note that this would not affect the Haskell daemons using cURL + GnuTLS,
    because we don't close all file descriptors on fork there.   

    Signed-off-by: Apollon Oikonomopoulos <apoikos@gmail.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

```
Originally added on 2014-10-06 09:26:52 +0000 UTC.

Changed State: Fixed
